[.NET] Dutch Merged support

LionbridgeCSII opened the pull request:
MergedExtractor: 86 pass, 0 fail.
MergedParser: 185 pass, 5 fail.

Failing cases:
- "Ik wil voor 3 mensen bij een pizzatent reserveren in seattle voor vanavond rond 20.00"
- "op vrijdag voor 3 's middags"
- "Zet de bruiloft van make cable in mijn agena voor woensdag de eenendertigste"
- "Zet de bruiloft van make cable in mijn agena voor dinsdag de eenendertigste"
- "Maak een afspraak voor Paasmaandag 2017"

These cases fail due to the ambiguity of "voor" that can be interpreted both as "for" and "before". 
Our idea would be to modify the test cases resolutions to match the results returned by the parser (as we did in Turkish for the ambiguity "in/within") because, since it seems not possible to resolve the ambiguity, the parser resolution can be considered correct according to one of the two interpretations. 
But we wanted to hear @iMicknl opinion before proceeding, in case he has some suggestion on how to handle these cases.

Some test cases produce different results respect to their English counterpart due to not completely faithful translations or different sentence constructions. In this case, we have updated the resolution of the test cases and added new inputs trying to better match the original meaning:
- "Ik zal morgen vertrekken" (I'll leave by tomorrow), expected daterange but returns date because "by" is missing ->  "Ik zal door morgen vertrekken".
- "Laten we over 3 minuten sinds vandaag beginnen" (Let's start 3 minutes from today), expected duration but returns datetime because "over" (in) is present -> "Laten we 3 minuten sinds vandaag beginnen".
Similarly for the cases "Laten we over 3 minuten sinds nu beginnen" and "Laten we over 3 minuten vanaf vandaag beginnen".
- "Het gebeurde op vorige week dinsdag om 2 uur 's middags" (It happened on previous Tuesday by 2 in the afternoon), expected datetimerange but returns datetime because "om" (at) is used instead of "by" -> "Het gebeurde op vorige week dinsdag tegen 2 uur 's middags".
- "Deze taak zal eerder dan 3 dagen na morgen af zijn" (This task will be done less than 3 days after tomorrow), the resolution is differente because "eerder" (earlier) is used instead of "less" -> "Deze taak zal minder dan 3 dagen na morgen af zijn".
- "pik Jordans medicijnen op bij de costco apotheek op havana ergens voor aankomende dinsdag om 12.00" (pick up jordan's medicine at the costco pharmacy on havana sometime before next tuesday at 12:00), the resolution is different because "aankomende" (upcoming) is used instead of "next" -> "pik Jordans medicijnen op bij de costco apotheek op havana ergens voor volgende dinsdag om 12.00".
- "Je kunt niet later dan 2016 vertrekken" (You can't leave later than 2016) returns a different result ("niet later dan 2016" instead of "later dan 2016") due to the different sentence construction. Similarly the case "Je kunt niet later dan of in 2016 vertrekken". We added a new test case "Ik zal uiterlijk morgen vertrekken" to show support for the analogous expression "uiterlijk" (no later than).
- "Ik zal binnen de komende dag en 5 uren terug zijn" (I will be back within the next 1 day and 5 hours), the missing "1" produces a wrong resolution (also in English) that needs to be fixed -> "Ik zal binnen de komende 1 dag en 5 uren terug zijn"

The following time expressions return double (ampm) resolution because, unlike their English counterparts, they do not disambiguate the two (am/pm) cases:
- "VOEG LUNCH TOE OM 12.30 OP VRIJ"
- "Voor vier maandag om 12.00"
- "overmorgen om 8.00"
- "leg een meeting vast iedere week om 8.00"

tellarin commented on Specs/DateTime/Dutch/MergedParser.json:
```diff
@@ -339,11 +352,12 @@
   },
   {
     "Input": "Zet de bruiloft van make cable in mijn agena voor woensdag de eenendertigste",
+	"Comment": "Ambiguity issue of 'voor' interpreted as 'before' when it should be interpreted as 'for'",
```
Nitpick: Indentation.

LionbridgeCSII replied:
fixed

tellarin commented:
@LionbridgeCSII Looks mostly good to me. But I'll wait for @iMicknl's feedback as he's a native speaker and also knows the project.

tellarin commented:
Build failed with a CI issue. Re-triggered build.

iMicknl commented:
> @LionbridgeCSII Looks mostly good to me. But I'll wait for @iMicknl's feedback as he's a native speaker and also knows the project.

@tellarin, I am not super familiar with the 'Merged' extractor/parser, what is the goal of this one? There are quite some testcases 

>"Ik zal morgen vertrekken" (I'll leave by tomorrow), expected daterange but returns date because "by" is missing -> "Ik zal door morgen vertrekken".

For example, the above sentence would be translated as 'Ik vertrek morgen', which is the same as 'I'm leaving tomorrow'. The proposed one with 'door morgen' is not a valid Dutch sentence.

LionbridgeCSII commented:
> For example, the above sentence would be translated as 'Ik vertrek morgen', which is the same as 'I'm leaving tomorrow'. The proposed one with 'door morgen' is not a valid Dutch sentence.

@iMicknl, the thing is that "I'm leaving tomorrow" and "I'm leaving by tomorrow" have different meanings in English and produce different results: "tomorrow" is a precise date, "by tomorrow" is like "before tomorrow" and is interpreted as daterange. 
So, we were trying to express the same meaning in Dutch (if possible). Maybe we could use "tegen" instead of "door" ("Ik zal tegen morgen vertrekken")? 
(I can find sentences which use it, e.g.  "Secnav wil die getuige hier, tegen morgen vroeg", "Regel dat de vaas en het boek hier zijn tegen morgen.", "Zeg hem dat ik die bewijzen nodig heb tegen morgen.")

iMicknl commented:
Good find, @LionbridgeCSII! I would agree that 'tegen morgen' is the best form to use when we look for a range from now until tomorrow. Some of the DateTime specs for Dutch are quite literally translated unfortunately, and I spotted indeed some mistakes.

If you are looking for examples for a daterange from moment of writing until tomorrow, perhaps these examples make sense as well? (but could be out of scope for this PR).

"Vergeet niet je huiswerk voor morgen in te leveren". 
"Je moet het boek uiterlijk morgen terugbrengen".

Will have a quick ook at the others asap.

iMicknl commented:
> The following time expressions return double (ampm) resolution because, unlike their English counterparts, they do not disambiguate the two (am/pm) cases:
> 
> * "VOEG LUNCH TOE OM 12.30 OP VRIJ"
> * "Voor vier maandag om 12.00"
> * "overmorgen om 8.00"
> * "leg een meeting vast iedere week om 8.00"

@tellarin in Dutch, 24 hour notation is commonly used, AM/PM is not used at all. Would it make sense to resolve them by  default based on 24 hour notation? However, this would only be the case for 8:00, since 'acht' (eight) cannot be resolved to a specific time. Is this implemented for other European languages? (otherwise please disregard this comment 👍🏻).

iMicknl commented on Specs/DateTime/Dutch/DatePeriodExtractor.json:
```diff
@@ -3075,14 +3075,14 @@
     ]
   },
   {
-    "Input": "Het zal 2 dagen vanaf nu gebeuren",
```
@LionbridgeCSII @tellarin this is the only case I am not certain about. In the English specs this is mentioned as `from now`, which would be in Dutch:

'vanaf nu' -> 'from now'
'vanaf vandaag' -> 'from today'

The current change with 'daarna' (afterwards, after that) is different. And for me, the issue would be that it is unclear to what it refers to. The original form mentions that explicit, where 'daarna' could be based on a previous sentence..

iMicknl replied:
For '2 days hence' I can understand this suggestion! 

https://github.com/microsoft/Recognizers-Text/blob/51ee58ab19624a6bf431fcdccc3779743b506d0f/Specs/DateTime/English/MergedParser.json#L3324

Perhaps it would make sense to add these cases, even if they are unsupported for now? Currently in English and Dutch only "3 minuten vanaf nu" is used, however this is not a test cases for days.

It would make sense to me to have this added as a testcase;
"Het evenement zal 2 dagen vanaf nu plaatsvinden."

LionbridgeCSII replied:
ok, I will add this test case.

LionbridgeCSII commented:
@iMicknl, could you give your opinion about these cases too?

> Failing cases:
> 
>     * "Ik wil voor 3 mensen bij een pizzatent reserveren in seattle voor vanavond rond 20.00"
> 
>     * "op vrijdag voor 3 's middags"
> 
>     * "Zet de bruiloft van make cable in mijn agena voor woensdag de eenendertigste"
> 
>     * "Zet de bruiloft van make cable in mijn agena voor dinsdag de eenendertigste"
> 
>     * "Maak een afspraak voor Paasmaandag 2017"
> 
> 
> These cases fail due to the ambiguity of "voor" that can be interpreted both as "for" and "before".
> Our idea would be to modify the test cases resolutions to match the results returned by the parser (as we did in Turkish for the ambiguity "in/within") because, since it seems not possible to resolve the ambiguity, the parser resolution can be considered correct according to one of the two interpretations.

LionbridgeCSII commented:
> > The following time expressions return double (ampm) resolution because, unlike their English counterparts, they do not disambiguate the two (am/pm) cases:
> > 
> > * "VOEG LUNCH TOE OM 12.30 OP VRIJ"
> > * "Voor vier maandag om 12.00"
> > * "overmorgen om 8.00"
> > * "leg een meeting vast iedere week om 8.00"
> 
> @tellarin in Dutch, 24 hour notation is commonly used, AM/PM is not used at all. Would it make sense to resolve them by default based on 24 hour notation? However, this would only be the case for 8:00, since 'acht' (eight) cannot be resolved to a specific time. Is this implemented for other European languages? (otherwise please disregard this comment 👍🏻).

@iMicknl, we considered this option but it gives more problems with other test cases. Also, it seems the 12-hour clock is used in Dutch in spoken language, so it seems better to keeps things as they are since test cases seem to refer to both written and spoken languages.

iMicknl commented:
> An even better solution could be to make the parsers return both interpretations. Maybe we should log an issue to address this as a possible enhancement, what do you think @tellarin?

I find this a hard one. You are correct, grammatically both could work. It really depends on the context. @tellarin what would be your preference and how is this handled for other (European) languages?

"Ik wil graag voor vanavond reserveren." (make a reservation for tonight). 
"Deze opdracht moet je voor vanavond inleveren" (you must submit this assignment before tonight)

I would say that the first one is the exception mainly, since other forms (with 'op', or without 'voor' are also commonly used), but in some situations it can definitely happen.

"Ik wil vanavond een tafel reserveren" or "Maak een afspraak op 3 mei".

> @iMicknl, we considered this option but it gives more problems with other test cases. Also, it seems the 12-hour clock is used in Dutch in spoken language, so it seems better to keeps things as they are since test cases refer to both written and spoken languages.

Makes sense, it could be logged as a future improvement, but I would say it is definitely out of scope for this one. I can imagine that this could be changed if many users face this issue eventually, @tellarin.

LionbridgeCSII commented:
@tellarin, given the above discussion, I have re-enabled the skipped test cases (modifying the resolution to fit the 'before' interpretation), replaced 'door morgen' with 'tegen morgen' and added the inputs suggested by @iMicknl.

tellarin commented on Specs/DateTime/Dutch/MergedParser.json:
```diff
@@ -120,19 +134,20 @@
   },
   {
     "Input": "Ik wil voor 3 mensen bij een pizzatent reserveren in seattle voor vanavond rond 20.00",
+	"Comment": "Ambiguity issue of 'voor' interpreted as 'before' when it should be interpreted as 'for'",
```
Nitpick: indentation

tellarin commented on Specs/DateTime/English/DateTimeModel.json:
```diff
@@ -20421,5 +20421,31 @@
+        "Resolution": {
+          "values": [
+            {
+              "timex": "(2016-11-07T16:12:00,2016-11-07T21:12:00,PT5H)",
```
This timex is not incorrect as the period only shows 5h.

LionbridgeCSII replied:
But it does not take into account the "next day" part and so the "end" turns out to be on the same day of the "start" while it should be on the next day. 
On the contrary, an expression like "within the next 1 day and 5 hours" returns the correct timex (2016-11-07T16:12:00,2016-11-08T21:12:00,P1DT5H)

tellarin replied:
Yeah... The timex generation/calculation is incorrect. But it's not really in the scope of this PR. Please open an issue to fix this and add it to the priority list. Thanks.

tellarin reviewed (changes requested):
@iMicknl, can you take a final look?
The flagged timex issue is independent of this PR purpose. It can be fixed as part of this PR or in a separate PR.

tellarin commented:
@LionbridgeCSII, do you have an ETA for the update to this PR? Thanks.

LionbridgeCSII commented:
> @LionbridgeCSII, do you have an ETA for the update to this PR? Thanks.

I have already updated this PR according to the comments from @iMicknl (https://github.com/microsoft/Recognizers-Text/pull/2589#issuecomment-846077178). Is there something else that need to be addressed? Otherwise I will only solve the conflicts.

tellarin commented:
> I have already updated this PR according to the comments from @iMicknl ([#2589 (comment)](https://github.com/microsoft/Recognizers-Text/pull/2589#issuecomment-846077178)). Is there something else that need to be addressed? Otherwise I will only solve the conflicts.

My bad. I thought there was still something to address.

LionbridgeCSII commented:
@tellarin, conflicts solved.

iMicknl commented on Patterns/Dutch/Dutch-DateTime.yaml:
```diff
@@ -1155,7 +1155,8 @@ HolidayNames: !dictionary
     christmas: [ kerstfeest, kerstmis, kerst, xmas ]
-    easterday: [ pasen, paasdag ]
+    easterday: [ pasen, paasdag, paaszondag ]
+    eastermonday: [ paasmaandag ]
```
Would be good to have `tweede paasdag` here as well, since this is more common.

LionbridgeCSII replied:
ok, added

iMicknl replied:
Thanks!

iMicknl commented on Patterns/Dutch/Dutch-DateTime.yaml:
```diff
@@ -1155,7 +1155,8 @@ HolidayNames: !dictionary
     whitesunday: [ pinksteren ]
     christmas: [ kerstfeest, kerstmis, kerst, xmas ]
-    easterday: [ pasen, paasdag ]
+    easterday: [ pasen, paasdag, paaszondag ]
```
`eerste paasdag` should be added, maybe even in favor of paasdag. There is a clear distinction in Dutch between 'eerste' and 'tweede' paasdag. Having paasdag as fallback to easterday is fine by me.

LionbridgeCSII replied:
added

iMicknl reviewed (approved):
Looks good to me! Just a minor remark around the added Holiday that I missed previously in this PR.

Not sure if it was intentionally added, or if it was a merge error.

tellarin reviewed (approved):
LGTM. Please resolve the merge conflict to merge.
Thanks!

LionbridgeCSII commented:
@tellarin, conflict solved
